staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,1036 @@
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,4,opt,name=nodeSelector"`
+
+	// AllNodes indicates that all nodes have access to the resources in the pool.
+	//
+	// Exactly one of NodeName, NodeSelector and AllNodes must be set.
+	//
+	// +optional
+	// +oneOf=NodeSelection
+	AllNodes bool `json:"allNodes,omitempty" protobuf:"bytes,5,opt,name=allNodes"`
+
+	// Devices lists some or all of the devices in this pool.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +optional
+	// +listType=atomic
+	Devices []Device `json:"devices" protobuf:"bytes,6,name=devices"`
+}
+
+// ResourcePool describes the pool that ResourceSlices belong to.
+type ResourcePool struct {
+	// Name is used to identify the pool. For node-local devices, this
+	// is often the node name, but this is not required.
+	//
+	// It must not be longer than 253 characters and must consist of one or more DNS sub-domains
+	// separated by slashes.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Generation tracks the change in a pool over time. Whenever a driver
+	// changes something about one or more of the resources in a pool, it
+	// must change the generation in all ResourceSlices which are part of
+	// that pool. Consumers of ResourceSlices should only consider
+	// resources from the pool with the highest generation number. The
+	// generation may be reset by drivers, which should be fine for
+	// consumers, assuming that all ResourceSlices in a pool are updated to
+	// match or deleted.
+	//
+	// Combined with ResourceSliceCount, this mechanism enables consumers to
+	// detect pools which are comprised of multiple ResourceSlices and are
+	// in an incomplete state.
+	//
+	// +required
+	Generation int64 `json:"generation" protobuf:"bytes,2,name=generation"`
+
+	// ResourceSliceCount is the total number of ResourceSlices in the pool at this
+	// generation number. Must be higher than zero.
+	//
+	// Consumers can use this to check whether they have seen all ResourceSlices
+	// belonging to the same pool.
+	//
+	// +required
+	ResourceSliceCount int64 `json:"resourceSliceCount" protobuf:"bytes,3,name=resourceSliceCount"`
+}
+
+const ResourceSliceMaxSharedCapacity = 128
+const ResourceSliceMaxDevices = 128
+const PoolNameMaxLength = validation.DNS1123SubdomainMaxLength // Same as for a single node name.
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Basic defines one device instance using fields that are supported
+	// by all clients which support DRA.
+	//
+	// +optional
+	// +oneOf=deviceType
+	Basic *BasicDevice `json:"basic,omitempty" protobuf:"bytes,2,opt,name=basic"`
```

Comment:
Let's start a thread about extending `BasicDevice` and `ResourceSlice` with shared attributes.

https://github.com/kubernetes/kubernetes/pull/125488#issuecomment-2226583572:
> > Do we think we want per-slice attribs in 1.31?
> I would say "yes", because it will usefully compact things, and can be ignored by the driver author if they don't need it. Should we do Capacity too?

It would be consistent.

About whether we should or should not do this in 1.31: if you are up to further API reviews, then I don't mind adding it. I'll finish the current round of updates and squash, then do shared attributes in a separate commit. We can then still decide whether we want them in this PR or a quick follow up PR.

I'm also tempted to do `DeviceRequest.OneOf` in 1.31. But first things first...

> This will cause one problem though. Since Name got bumped up out of BasicDevice, all there is in BasicDevice are attributes and capacities. If a device doesn't have any per-device attribute or capacity, except for the name, then BasicDevice will be empty.

As [Tim said](https://github.com/kubernetes/kubernetes/pull/125488#issuecomment-2226740829), that should be okay. The `omitempty` is for the pointer, not the content of the struct.

Reply:
If Tim is OK with adding the shared attrs/capacity in this PR, I am good with it too. If we add `DeviceRequest.OneOf`, it should be a separate PR. I don't want to add risk that this one misses. In fact I am guessing it would be in 1.32 instead, seems most likely. We'll have to decide on a bunch of additions in 1.32 alpha, we'll have to decide how to organize them (how many KEPs, etc.).

Reply:
>  if you are up to further API reviews

Hopefully this is easy from the API POV

> do shared attributes in a separate commit. 

THANK YOU

Reply:
I would vote *not* to add these in 1.31. 

The reason being that at least one of the partitionable devices alternatives (my last one that doesn't have a prototype yet), would argue for *not* having shared attributes defined at the slice level. Instead there would be a separate object for defining a "device shape", and this is where the ability to have shared attributes would exist.

If we add them both there and in the slice itself, we have to do some mental gymnastics to reason about what set of attributes take precedence. Meaning that if we go this route, I'd rather just give "device shapes" the ability to inherit attributes from a global set, but not the devices themselves.

I realize its unclear whether we will go the route of "device shapes" in the future, but I'd rather not introduce shared attributes now, when it will make our lives harder later.

Reply:
If we DID add them, they would have to go in `spec`, right?  Meaning that they apply to devices and gizmos and so on.  If we only want them to apply to devices, then we sould need to add ANOTHER indirection:


```diff
 	// Devices lists some or all of the devices in this pool.
-	//
-	// Must not have more than 128 entries.
 	//
 	// +optional
-	// +listType=atomic
- 	Devices []Device `json:"devices" protobuf:"bytes,6,name=devices"`
+ 	Devices SpecDevices `json:"devices" protobuf:"bytes,6,name=devices"`
```

and then

```
type SpecDevices struct {
    // Attributes defines shared attributs for all devices.
	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,1,rep,name=attributes"`

	// Devices lists some or all of the devices in this pool.
	//
	// Must not have more than 128 entries.
	//
	// +optional
	// +listType=atomic
	Devices []Device `json:"devices" protobuf:"bytes,2,name=devices"`
```

I suspect we are not keen on that, but if we are, we need to do it now.  I am OK to hold off on shared attribs, I just don't want to "forget" and then lose by default.  Not having them from the beginning will make introducing them a bit more dangerous for users, which is why I have them on my to-do list.

Reply:
We could add `slice.spec.defaultDeviceAttributes` alongside `slice.spec.devices[*]`. Might be better then nesting `devices` inside `devices`.

But with the "device shape" objects proposed by @klueska we don't need these `defaultDeviceAttributes` at all. Instead, individual devices explicitly specify what they inherit. This seems reasonable to me.

As there's no clear consensus, I prefer to not try some last minute updates.